src/UpgradeRule/AddNamespaceRule.php

```diff
@@ -77,6 +79,43 @@ public function withRoot($root)
         return $this;
     }
 
+
+    public function beforeUpgrade(CollectionInterface $code, CodeChangeSet $changeset)
+    {
+        // Before upgrading, collect all "to be namespaced" classes
+        foreach ($code->iterateItems() as $file) {
+            $this->beforeUpgradeFile($file);
+        }
+    }
+
+    /**
+     * Find and record all classes in this file
+     *
+     * @param ItemInterface $file
+     */
+    protected function beforeUpgradeFile(ItemInterface $file)
+    {
+        if (!$this->appliesTo($file)) {
+            return;
+        }
+
+        // Get classes in this file
+        $contents = $file->getContents();
+        $source = new MutableSource($contents);
+        list($fileNamespace, $fileClasses) = $this->findClasses($source);
+
+        // Skip if file already namespaced
+        if ($fileNamespace) {
+            return;
+        }
+
+        // Add classes in this file to namespace mapping
+        $newNamespace = $this->getNamespaceForFile($file);
+        foreach ($fileClasses as $class) {
+            $this->renamedClasses[$class] = "{$newNamespace}\\{$class}";
+        }
+    }
+
     public function upgradeFile($contents, $file)
```

Comment:
We want to amend the call approach here to more closely match the before handlers. Maybe pass an ItemInterface and a CodeChangeSet to populate?